hypothesis-python/src/hypothesis/provisional.py

```diff
@@ -20,62 +20,159 @@
 It is intended for internal use, to ease code reuse, and is not stable.
 Point releases may move or break the contents at any time!
 
-Internet strategies should conform to https://tools.ietf.org/html/rfc3696 or
-the authoritative definitions it links to.  If not, report the bug!
+Internet strategies should conform to :rfc:`3986` or the authoritative
+definitions it links to.  If not, report the bug!
 """
+# https://tools.ietf.org/html/rfc3696
 
 from __future__ import absolute_import, division, print_function
 
+import os.path
 import string
 
 import hypothesis._strategies as st
+import hypothesis.internal.conjecture.utils as cu
+from hypothesis.errors import InvalidArgument
+from hypothesis.searchstrategy.strategies import SearchStrategy
+
+if False:
+    from typing import Text  # noqa
+    from hypothesis.searchstrategy.strategies import SearchStrategy, Ex  # noqa
+
+
+URL_SAFE_CHARACTERS = set(string.ascii_letters + string.digits + "$-_.+!*'(),")
+
+
+# This file is sourced from http://data.iana.org/TLD/tlds-alpha-by-domain.txt
+# The file contains additional information about the date that it was last updated.
+with open(
+    os.path.join(os.path.dirname(__file__), "vendor", "tlds-alpha-by-domain.txt")
+) as tld_file:
+    __header = next(tld_file)
+    assert __header.startswith("#")
+    TOP_LEVEL_DOMAINS = sorted((line.rstrip() for line in tld_file), key=len)
+TOP_LEVEL_DOMAINS.insert(0, "COM")
+
+
+class DomainNameStrategy(SearchStrategy):
+    @staticmethod
+    def clean_inputs(minimum, maximum, value, variable_name):
+        if value is None:
+            value = maximum
+        elif not isinstance(value, int):
+            raise InvalidArgument(
+                "Expected integer but %s is a %s"
+                % (variable_name, type(value).__name__)
+            )
+        elif not minimum <= value <= maximum:
+            raise InvalidArgument(
+                "Invalid value %r < %s=%r < %r"
+                % (minimum, variable_name, value, maximum)
+            )
+        return value
+
+    def __init__(self, max_length=None, max_element_length=None):
+        """
+        A strategy for :rfc:`1035` fully qualified domain names.
+
+        The upper limit for max_length is 255 in accordance with :rfc:`1035#section-2.3.4`
+        The lower limit for max_length is 4, corresponding to a two letter domain
+        with a single letter sub-domain.
+        The upper limit for max_element_length is 63 in accordance with :rfc:`1035#section-2.3.4`
+        The lower limit for max_element_length is 1 in accordance with :rfc:`1035#section-2.3.4`
+        """
+        # https://tools.ietf.org/html/rfc1035#section-2.3.4
+
+        max_length = self.clean_inputs(4, 255, max_length, "max_length")
+        max_element_length = self.clean_inputs(
+            1, 63, max_element_length, "max_element_length"
+        )
+
+        SearchStrategy.__init__(self)
```

Comment:
My usual inclination is to use `super()` to call the parent class constructor, i.e. `super(DomainNameStrategy, self).__init__()` – but I can’t off the top of my head remember the benefits.

Reply:
I've looked at the other classes, and switched to using `super(DomainNameStrategy, self).__init__()` since thats the pattern used elsewhere.